stan/math/rev/core/grad.hpp

```diff
@@ -28,20 +28,11 @@ namespace math {
  * derivative propagation.
  */
 static void grad(vari* vi) {
-  // simple reference implementation (intended as doc):
-  //   vi->init_dependent();
-  //   size_t end = var_stack_.size();
-  //   size_t begin = empty_nested() ? 0 : end - nested_size();
-  //   for (size_t i = end; --i > begin; )
-  //     var_stack_[i]->chain();
-
-  using it_t = std::vector<vari*>::reverse_iterator;
   vi->init_dependent();
-  it_t begin = ChainableStack::instance_->var_stack_.rbegin();
-  it_t end = empty_nested() ? ChainableStack::instance_->var_stack_.rend()
-                            : begin + nested_size();
-  for (it_t it = begin; it < end; ++it) {
-    (*it)->chain();
+  size_t end = ChainableStack::instance_->var_stack_.size();
+  size_t begin = empty_nested() ? 0 : end - nested_size();
+  for (size_t i = end; i > begin; --i) {
```

Comment:
There's an idiom for these loops counting downward with `size_t`:

```
for (size_t i = end; --i > begin; ) {
  foo(...var_stack[i]...)
}
```

It works in the case that `begin = 0` with `size_t` variables.

Reply:
I switched back to this (that's how it was originally).

Reply:
I undid this change.

Correct version was:
```
for (size_t i = end; i-- > begin; ) {
  foo(...var_stack[i]...)
}
```

And I figure if we're making mistakes with the difference between --i and i--, we should favor a different solution.